source/server/hot_restarting_base.cc

```diff
@@ -41,77 +43,171 @@ void HotRestartingBase::bindDomainSocket(uint64_t id, const std::string& role) {
+
+    out->mutable_pass_listen_socket_reply()->set_fd(*reinterpret_cast<int*>(CMSG_DATA(cmsg)));
+
+    RELEASE_ASSERT(CMSG_NXTHDR(message, cmsg) == nullptr,
+                   "More than one control data on a single hot restart recvmsg().");
+  }
+}
+
+std::unique_ptr<HotRestartMessage> HotRestartingBase::receiveHotRestartMessage(bool block) {
   // By default the domain socket is non blocking. If we need to block, make it blocking first.
   if (block) {
-    int rc = fcntl(my_domain_socket_, F_SETFL, 0);
-    RELEASE_ASSERT(rc != -1, "");
+    RELEASE_ASSERT(fcntl(my_domain_socket_, F_SETFL, 0) != -1,
+                   fmt::format("Set domain socket blocking failed, errno = {}", errno));
+  }
+  if (recv_buf_.size() < MaxSendmsgSize) {
+    resizeRecvBuf(MaxSendmsgSize, 0);
   }
 
   iovec iov[1];
-  iov[0].iov_base = &rpc_buffer_[0];
-  iov[0].iov_len = rpc_buffer_.size();
-
-  // We always setup to receive an FD even though most messages do not pass one.
-  uint8_t control_buffer[CMSG_SPACE(sizeof(int))];
-  memset(control_buffer, 0, CMSG_SPACE(sizeof(int)));
-
   msghdr message;
-  memset(&message, 0, sizeof(message));
-  message.msg_iov = iov;
-  message.msg_iovlen = 1;
-  message.msg_control = control_buffer;
-  message.msg_controllen = CMSG_SPACE(sizeof(int));
-
-  int rc = recvmsg(my_domain_socket_, &message, 0);
-  if (!block && rc == -1 && errno == EAGAIN) {
-    return nullptr;
+  uint8_t control_buffer[CMSG_SPACE(sizeof(int))];
+  std::unique_ptr<HotRestartMessage> ret = nullptr;
+  while (!ret) {
+    iov[0].iov_base = receive_next_byte_here_;
+    iov[0].iov_len = MaxSendmsgSize;
+
+    // We always setup to receive an FD even though most messages do not pass one.
+    memset(control_buffer, 0, CMSG_SPACE(sizeof(int)));
+    memset(&message, 0, sizeof(message));
+    message.msg_iov = iov;
+    message.msg_iovlen = 1;
+    message.msg_control = control_buffer;
+    message.msg_controllen = CMSG_SPACE(sizeof(int));
+
+    int recvmsg_rc = recvmsg(my_domain_socket_, &message, 0);
+    if (!block && recvmsg_rc == -1 && errno == EAGAIN) {
+      return nullptr;
+    }
+    RELEASE_ASSERT(recvmsg_rc != -1, fmt::format("recvmsg() returned -1, errno = {}", errno));
+    RELEASE_ASSERT(message.msg_flags == 0,
+                   fmt::format("recvmsg() left msg_flags = {}", message.msg_flags));
+    receive_next_byte_here_ += recvmsg_rc;
+
+    // If we're in the middle of receiving a protobuf, the whole recvmsg() was protobuf data.
+    // Otherwise, the first 8 bytes was 'length'.
+    partial_proto_bytes_ +=
+        expected_message_length_.has_value() ? recvmsg_rc : recvmsg_rc - sizeof(uint64_t);
+
+    // If we don't already know 'length', we're at the start of a new length+protobuf message!
+    if (!expected_message_length_.has_value()) {
+      // We are not ok with messages so fragmented that the length doesn't even come in one piece.
+      RELEASE_ASSERT(recvmsg_rc >= 8, "received a brokenly tiny message fragment.");
```

Comment:
nit: in some places you use 8 and in some places you use sizeof(uint64_t). Can we do the latter everywhere?

Reply:
Looks like I left one other '8' in; fixed that one. However, this one is to avoid the compiler complaining about a signed/unsigned comparison.